File: gachisa-backend/src/main/java/com/gachisa/payment/service/dto/RefundPreparation.java
package com.gachisa.payment.service.dto;

import com.gachisa.payment.entity.PaymentAttempt;
import com.gachisa.payment.entity.Refund;

public record RefundPreparation(
        Long refundId,
        String paymentKey,
        String reason,
        String pgIdempotencyKey,
        boolean requestRequired
) {
    public static RefundPreparation request(Refund refund, PaymentAttempt attempt) {
        return new RefundPreparation(
                refund.getId(), attempt.getPgPaymentKey(), refund.getReason(),
                refund.getPgIdempotencyKey(), true);
    }

    public static RefundPreparation existing(Refund refund) {
        return new RefundPreparation(
                refund.getId(), null, refund.getReason(), refund.getPgIdempotencyKey(), false);
    }
}

File: gachisa-backend/src/main/kotlin/com/gachisa/payment/service/dto/RefundPreparation.kt
package com.gachisa.payment.service.dto

import com.gachisa.payment.entity.PaymentAttempt
import com.gachisa.payment.entity.Refund

data class RefundPreparation(
    val refundId: Long,
    val paymentKey: String?,
    val reason: String,
    val pgIdempotencyKey: String,
    val requestRequired: Boolean,
) {
    companion object {
        fun request(refund: Refund, attempt: PaymentAttempt): RefundPreparation = RefundPreparation(
            refundId = refund.id!!,
            paymentKey = attempt.pgPaymentKey,
            reason = refund.reason,
            pgIdempotencyKey = refund.pgIdempotencyKey,
            requestRequired = true,
        )

        fun existing(refund: Refund): RefundPreparation = RefundPreparation(
            refundId = refund.id!!,
            paymentKey = null,
            reason = refund.reason,
            pgIdempotencyKey = refund.pgIdempotencyKey,
            requestRequired = false,
        )
    }
}
